### PR TITLE
upgpkg: miopen-hip/opencl 5.1.0-1

### PR DESCRIPTION
**Pull request by JP-Ellis**
There a few notable changes here:

- Instead of building Boost, the system versions are being used. This has become   possible since boostorg/config#393 which is included in Boost 1.78. The fix   relies in `HIP_VERSION` being defined which unfortunately is not always the case (and I am as yet unsure why). As a temporary workaround, the nine   problematic files are manually patched with a `#include <hip/hip_version.h>`. This is not an issue with the OpenCL version.
- I have similarly built against the system version of Half and SQLite. It appears there are no issues due to this.
- I have also tried removing the dependency `ROCmSoftwarePlatform/llvm-project-mlir` and everything compiles fine. I am unsure whether other ROCm packages provide what is necessary or not.
- I have been encountering an issue when compiling `build/kernel.cpp` with the HIP compiler. I do not know why, but disabling optimizations for that file works.

I will mark this pull request as a draft for now until the points above can be discussed.  Update: I am unable to mark it as a draft... someone with write access to rocm-arch needs to do it.

Patches: #676

### CODE REVIEW

**Comment by JP-Ellis** [comment]
Has anyone else had a chance to look into the above changes?  Did we want to merge this, or wait until upstream fixes these issues?

**Comment by bernharl** [comment]
Does upstream also build ´build/kernel.cpp´ with O0?

**Comment by JP-Ellis** [comment]
I have created an upstream issue with regards to the missing `hip_version.h` header at ROCmSoftwarePlatform/MIOpen#1489.  I'll have another look at the compilation seg faults for `kernel.cpp`.

**Comment by JP-Ellis** [comment]
With regards to `build/kernel.cpp`, I had no issue building it when creating the PR for the missing `HIP_VERSION`.  It would appear like this is fixed upstream, though I need to find out which commit fixed the issue so that I can include it as a patch.

**Comment by JP-Ellis** [comment]
I have updated the PR to now to work with version 5.1.0.  Some of the new changes:

- Instead of building Boost, the system versions are being used. This has become possible since boostorg/config#393 which is included in Boost 1.78. The fix relies in `HIP_VERSION` being defined which unfortunately is not always the case. I have submitted an upstream PR (see ROCmSoftwarePlatform/MIOpen#1389 and ROCmSoftwarePlatform/MIOpen#1490) which is included in the PKGBUILD.
- I have similarly built against the system version of SQLite. It appears there are no issues due to this.
- Instead of downloading and building LLVM MLIR every time, I have split that into a separate package now that it has a 5.1 release.  I have called this package `rocm-llvm-mlir` though feel free to suggest another package name.

File: rocm-llvm-mlir/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Maintainer: JP-Ellis <josh@jpellis.me>
+
```
Comment (tpkessler) [inline rocm-llvm-mlir/PKGBUILD:2]: Do you mind adding me as co-maintainer as I will be pushing it to the AUR?

Comment (JP-Ellis) [reply rocm-llvm-mlir/PKGBUILD:2]: Sure 😁

**Comment by wuxxin** [comment]
i tried the attached PKGBUILD for rocm-llvm-mlir and miopen-hip, so far it compiles on my machine, next step will be building pytorch-rocm for further testing of miopen-hip. ty for your work

**Comment by tpkessler** [comment]
Thank you very much @JP-Ellis for your work on updating these packages!

**Pull request merged**